src/commands/cmd_bulk_insert.c

```diff
@@ -83,24 +83,28 @@ void _MGraph_BulkInsert(void *args) {
         if (gc == NULL) {
             RedisModule_ReplyWithError(ctx, "Failed to allocate space for graph.");
             goto cleanup;
-        } else if (argc == 0) {
+        }
+        // Lock the graph for writing.
+        Graph_AcquireWriteLock(gc->g);
+
+        if (argc == 0) {
             // Only the allocation string was received, so our work is done.
             char success[1024] = {0};
             int len = snprintf(success, 1024, "Initialized a graph to accommodate %lld nodes and %lld edges.", final_node_count, final_edge_count);
             RedisModule_ReplyWithStringBuffer(ctx, success, len);
             goto cleanup;
         }
     } else {
+        // Query did not start with a "BEGIN" token
         gc = GraphContext_Retrieve(ctx, rs_graph_name);
         if (gc == NULL) {
             RedisModule_ReplyWithError(ctx, "Bulk insert query did not include a BEGIN token and graph was not found.");
             goto cleanup;
         }
+        // Lock the graph for writing.
+        Graph_AcquireWriteLock(gc->g);
```

Comment:
I think it would be possible to combine Graph_AcquireWriteLock into a single call right before the call to Graph_SetMatrixPolicy(gc->g, RESIZE_TO_CAPACITY);

Reply:
That was the original setup, but if the conditional on 90 triggers (we allocated a graph and have no further work), we still need to be holding a lock, as we jump straight to cleanup in that case and the unlock there is invalid.